Define EncodedClientHelloInner before outer extensions. (#503)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

does not include the Handshake structure's four byte header. The `zeros` field
MUST be all zeroes.

Repeating large extensions, such as "key_share" with post-quantum algorithms,
between ClientHelloInner and ClientHelloOuter can lead to excessive size. To

> reduce the size impact, the client then MAY substitute extensions which it knows

Reviewer: ```suggestion
reduce the size impact, the client MAY substitute extensions which it knows
```

will be duplicated in ClientHelloOuter. It does so by removing and replacing
extensions from EncodedClientHelloInner with a single "ech_outer_extensions"
extension, defined as follows:

~~~
    enum {
       ech_outer_extensions(0xfd00), (65535)
    } ExtensionType;

    ExtensionType OuterExtensions<2..254>;
~~~

OuterExtensions contains the removed ExtensionType values. Each value references
the matching extension in ClientHelloOuter. The values MUST be ordered
contiguously in ClientHelloInner, and the "ech_outer_extensions" extension MUST
be inserted in the corresponding position in EncodedClientHelloInner.
Additionally, the extensions MUST appear in ClientHelloOuter in the same

> relative order, however, there is no requirement that they be continguous. For

Reviewer: ```suggestion
relative order. However, there is no requirement that they be contiguous. For
```

example, OuterExtensions may contain extensions A, B, C, while ClientHelloOuter
contains extensions A, D, B, C, E, F.

The "ech_outer_extensions" extension can only be included in

> EncodedClientHelloInner, and MUST NOT be sent in either ClientHelloOuter or

Reviewer: ```suggestion
EncodedClientHelloInner, and MUST NOT appear in either ClientHelloOuter or
```
